src/qibo/parameter.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import sympy as sp
+
+from qibo.config import raise_error
+
+
+def calculate_derivatives(func):
+    """Calculates derivatives w.r.t to all parameters of a target function `func`."""
+    vars = []
+    for i in range(func.__code__.co_argcount):
+        vars.append(sp.Symbol(f"p{i}"))
+
+    expr = sp.sympify(func(*vars))
+
+    derivatives = []
+    for i in range(len(vars)):
+        derivative_expr = sp.diff(expr, vars[i])
+        derivatives.append(sp.lambdify(vars, derivative_expr))
+
+    return derivatives
+
+
+class Parameter:
+    """Object which allows for variational gate parameters. Several trainable parameters
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter. All possible analytical derivatives of the lambda function are
+    calculated at the object initialisation using Sympy.
+
+    Example::
+
+        from qibo.parameter import Parameter
+        param = Parameter(
+                lambda x, th1, th2, th3: x**2 * th1 + th2 * th3**2,
+                features=[7.0],
+                trainable=[1.5, 2.0, 3.0],
+            )
+
+        partial_derivative = param.get_partial_derivative(3)
+
+        param.update_parameters(trainable=[15.0, 10.0, 7.0], feature=[5.0])
+        param_value = param()
+
+
+    Args:
+        func (function): lambda function which builds the gate parameter. If both features and trainable parameters
+        compose the function, it must be passed by first providing the features and then the parameters, as
+        described in the code example above.
+        features (list or np.ndarray): array containing possible input features x.
+        trainable (list or np.ndarray): array with initial trainable parameters theta.
+        nofeatures (bool): flag to explicitly ban the updating of the features. This simplifies the task of updating Parameter objects simultaneously when some have embedded features and some do not.
+    """
+
+    def __init__(self, func, trainable=[], features=[]):
+        self.trainable = trainable
+        self.features = features
+
+        if self.nfeat + self.nparams != func.__code__.co_argcount:
+            raise_error(
+                TypeError,
+                f"{self.nfeat + self.nparams} parameters are provided, but {func.__code__.co_argcount} are required, please initialize features and trainable according to the defined function.",
+            )
```

Comment:
I only wanted to be more explicit in the error raising: I would like to underline the importance of setting both trainable and features.

Reply:
I would put the emphasis in the docs, and catching the error where it will be generating (if you want to raise a more informative one, but the one are actually getting could be even already informative enough).

Reply:
I think we have some ambiguities if we don't do like this. Let's suppose we define:

```python
param = Parameter(
           lambda x1, x2, th1, th2, th3: x1**2 * th1 + x2 * th2 * th3,
           trainable=[1.5, 2.0, 3.0],
       )
```

This is still be initialised as it is. But then, once one try to call `param()`, the raised error is:

```sh
TypeError: <lambda>() missing 2 required positional arguments: 'th2' and 'th3'
```

because features are appended before trainables into the components list.
I feel this ambiguous in the sense that `th2` and `th3` are here initialized.
My purposed error is more general and less confusing (my feeling).